src/lib/cid.test.js

```diff
@@ -0,0 +1,166 @@
+/* global describe it expect */
+// @ts-check
+import { toCidOrNull, getCodecOrNull, getCodeOrNull, toCidStrOrNull } from './cid'
+import * as multiformats from 'multiformats'
+import * as dagCbor from '@ipld/dag-cbor'
+import * as dagPb from '@ipld/dag-pb'
+import crypto from 'crypto'
+
+const { CID, hasher, bytes } = multiformats
+
+export const sha256 = hasher.from({
+  name: 'sha2-256',
+  code: 0x12,
+  encode: (input) => bytes.coerce(crypto.createHash('sha256').update(input).digest())
+})
+/**
+ *
+ * @template {number} Code
+ * @param {any} value
+ * @param {import('multiformats/block').BlockEncoder<Code, any>} codec
+ * @param {import('multiformats/hashes/interface').MultihashHasher<Code>} hasher
+ * @returns
+ */
+const createCID = async (value, codec, hasher) => {
+  try {
+    const digest = await hasher.digest(codec.encode(value))
+    return CID.create(1, codec.code, digest)
+  } catch (err) {
+    console.log('Failed to create CID', value, err)
+    return null
+  }
+}
+
+// base58btc - cidv1 - dag-cbor - sha2-256~256~63C300F377227B01B45396434D0AB912F2511A09BDFFFD61CB06E9765F76BFE8)
+// const cidStrOld = 'zdpuAs8sJjcmsPUfB1bUViftCZ8usnvs2cXrPH6MDyT4zrvSs'
+// base32 - cidv1 - dag-cbor - sha2-256~256~63C300F377227B01B45396434D0AB912F2511A09BDFFFD61CB06E9765F76BFE8)
+const cidStr = 'bafyreiddymapg5zcpma3iu4wingqvois6jirucn5776wdsyg5f3f65v75a'
```

Comment:
@rvagg One question that is still open is why these CIDs are changing. If the tests are passed the original CID, 'zdpuAs8sJjcmsPUfB1bUViftCZ8usnvs2cXrPH6MDyT4zrvSs', it fails expectations because the libraries are translating it to 'bafyreiddymapg5zcpma3iu4wingqvois6jirucn5776wdsyg5f3f65v75a'. When I pass the expected CID to it, the tests are passing correctly. 

I assume this is related to `base58btc` no longer being a default or something similar, but it still seems unexpected.

Reply:
_(I hope all the text below is helpful, I know it's long!)_

It's related to this: https://github.com/multiformats/js-cid/blob/2ed9449c7a7d2df522485822ae46f2d8d10fbbcc/src/index.js#L268-L273

Previously, when you decoded a CID by its multibase form it'd cache both the multibase you passed in to it _plus_ the string form—so you get `cid.multibaseName` and `cid.string` as cached values.

We removed that from CID, along with knowledge of all but the most fundamental base encodings (base58btc and base32) and now default to base32 for CIDv1 and base58btc for CIDv0 and _that's it_. If you want another base, then you need to bring along your own base encoder (that also keeps bundles small if you never have a use for novel bases). There's minimal magic and caching either. A CID is now just an encapsulation of the bytes of a CID, with some utility properties and methods to access and format it nicely. This is how it should have been but a lot of fluff was added to js-cid to make it fast and overly functional.

So the right way to view `zdpuAs8sJjcmsPUfB1bUViftCZ8usnvs2cXrPH6MDyT4zrvSs` and `bafyreiddymapg5zcpma3iu4wingqvois6jirucn5776wdsyg5f3f65v75a` are that they are just multibase encoded forms of bytes that represent a CID and in fact represent the exact same CID. They both represent the bytes `0x0171122063c300f377227b01b45396434d0ab912f2511a09bdfffd61cb06e9765f76bfe8`. Use js-multiformats to get access to those bytes, wrapped in a `CID` object. Any information about the base encoding is incidental. If you want to preserve base in a round-trip then that becomes an external concern (I guess you could argue that's an inconvenience that we should provide affordances for, but someone would have to make that case and so far it hasn't shown up). Previously, with js-cid, these were "equal" only in the sense that `equals()` would tell you they were. With js-multiformats, the intention is that all of the public properties are equal when you decode from either of these strings, because they are the same CID, just printed differently (as if you chose a different font to print the same word).

So back to failing assertions - which ones are failing if you use the `zdpu` form? I think the only ones that should be concern are the ones involving `toCidStrOrNull`, which is doing a `cid.toString()` and because the input base is not cached, it's coming out as the default base. What you could do is keep the `zdpu` form but any time a test does a round-trip through `toCidStrOrNull` you assert it against the `bafy` form, _or_ you provide a way to pass in the base58btc encoder to get that out.

Do you know if this impacts the application? Does it need to round-trip and preserve base anywhere in the explorer UI? If so, then we may need to build something in here to detect and preserve the base as it comes in and out. We could import basics.js and grab all of the bases from there, and have a map of the prefixes and store `input[0]` somewhere to preserve it. It would be interesting to know if the use-case here has trouble with the more minimal js-multiformats CID interface.

Reply:
I don't know if this impacts the application. I'm still trying to grasp all of the implications of IPLD and CIDs. One interesting thing regarding the older CIDs is that without my PR changes, a base32 CID seems to show up incorrectly

### `zb2rhZMC2PFynWT7oBj7e6BpDpzge367etSQi6ZUA81EVVCxG`

https://storybookipldexploretech.on.fleek.co/?path=/story/cid-info--cid-v1-b32-raw - shows that it's base58btc in the explorer (I just integrated fleek.co so that our storybook stories are published from master)
https://cid.ipfs.tech/#zb2rhZMC2PFynWT7oBj7e6BpDpzge367etSQi6ZUA81EVVCxG - shows the same thing as storybook
https://explore.ipld.io/#/explore/zb2rhZMC2PFynWT7oBj7e6BpDpzge367etSQi6ZUA81EVVCxG - doesn't load

### cidv1 version, `bafkreibigjl4qgeryaojdy7q5gtztodvivjwkiwog2p7p73h5obeakoi2e`

https://cid.ipfs.tech/#bafkreibigjl4qgeryaojdy7q5gtztodvivjwkiwog2p7p73h5obeakoi2e - shows the expected base
https://explore.ipld.io/#/explore/bafkreibigjl4qgeryaojdy7q5gtztodvivjwkiwog2p7p73h5obeakoi2e - doesn't load